data/transactions/verify/txn.go

```diff
@@ -448,3 +476,373 @@ func (w *worksetBuilder) next() (txnGroups [][]transactions.SignedTxn) {
+type VerificationResult struct {
+	TxnGroup       []transactions.SignedTxn
+	BacklogMessage interface{}
+	Err            error
+}
+
+// StreamVerifier verifies txn groups received through the stxnChan channel, and returns the
+// results through the resultChan
+type StreamVerifier struct {
+	resultChan       chan<- VerificationResult
+	stxnChan         <-chan UnverifiedElement
+	verificationPool execpool.BacklogPool
+	ctx              context.Context
+	cache            VerifiedTransactionCache
+	activeLoopWg     sync.WaitGroup
+	nbw              *NewBlockWatcher
+	ledger           logic.LedgerForSignature
+}
+
+// NewBlockWatcher is a struct used to provide a new block header to the
+// stream verifier
+type NewBlockWatcher struct {
+	blkHeader bookkeeping.BlockHeader
+	mu        deadlock.RWMutex
+}
+
+// MakeNewBlockWatcher construct a new block watcher with the initial blkHdr
+func MakeNewBlockWatcher(blkHdr bookkeeping.BlockHeader) (nbw *NewBlockWatcher) {
+	nbw = &NewBlockWatcher{
+		blkHeader: blkHdr,
+	}
+	return nbw
+}
+
+// OnNewBlock implements the interface to subscribe to new block notifications from the ledger
+func (nbw *NewBlockWatcher) OnNewBlock(block bookkeeping.Block, delta ledgercore.StateDelta) {
+	if nbw.blkHeader.Round >= block.BlockHeader.Round {
+		return
+	}
+	nbw.mu.Lock()
+	defer nbw.mu.Unlock()
+	nbw.blkHeader = block.BlockHeader
+}
+
+func (nbw *NewBlockWatcher) getBlockHeader() (bh *bookkeeping.BlockHeader) {
+	nbw.mu.RLock()
+	defer nbw.mu.RUnlock()
+	return &nbw.blkHeader
+}
+
+type batchLoad struct {
+	txnGroups             [][]transactions.SignedTxn
+	groupCtxs             []*GroupContext
+	elementBacklogMessage []interface{}
+	messagesForTxn        []int
+}
+
+func makeBatchLoad() (bl batchLoad) {
+	bl.txnGroups = make([][]transactions.SignedTxn, 0)
+	bl.groupCtxs = make([]*GroupContext, 0)
+	bl.elementBacklogMessage = make([]interface{}, 0)
+	bl.messagesForTxn = make([]int, 0)
+	return bl
+}
+
+func (bl *batchLoad) addLoad(txngrp []transactions.SignedTxn, gctx *GroupContext, backlogMsg interface{}, numBatchableSigs int) {
+	bl.txnGroups = append(bl.txnGroups, txngrp)
+	bl.groupCtxs = append(bl.groupCtxs, gctx)
+	bl.elementBacklogMessage = append(bl.elementBacklogMessage, backlogMsg)
+	bl.messagesForTxn = append(bl.messagesForTxn, numBatchableSigs)
+
+}
+
+// MakeStreamVerifier creates a new stream verifier and returns the chans used to send txn groups
+// to it and obtain the txn signature verification result from
+func MakeStreamVerifier(ctx context.Context, stxnChan <-chan UnverifiedElement, resultChan chan<- VerificationResult,
+	ledger logic.LedgerForSignature, nbw *NewBlockWatcher, verificationPool execpool.BacklogPool,
+	cache VerifiedTransactionCache) (sv *StreamVerifier) {
+
+	sv = &StreamVerifier{
+		resultChan:       resultChan,
+		stxnChan:         stxnChan,
+		verificationPool: verificationPool,
+		ctx:              ctx,
+		cache:            cache,
+		nbw:              nbw,
```

Comment:
I need a specific OnNewBlock implementation here with it's own lock.

Reply:
Instead of 
https://github.com/algonautshant/go-algorand/blob/f83101b839c9cb0c1d9439bdb90dd6ecce032c6f/data/transactions/verify/txn.go#L802
why not just use
```
	r := sv.ledger.Latest()
	last, err := node.ledger.Block(r)
```

Reply:
This pair of operations need to be done all the time, and this is an expensive operation on the ledger. 
It is much cheaper to get a local struct updated whenever a new block is created.